Take lock before CredWrite to avoid memory corruptions

juturu opened the pull request.

juturu committed:
creating std::string in setPassword

juturu committed:
Take lock before CredWrite to avoid memory corruptions

chandrachivukula commented on src/keytar_win.cc:
```diff
@@ -19,6 +20,8 @@ KEYTAR_OP_RESULT SetPassword(const std::string& service,
                  std::string* errStr) {
   std::string target_name = service + '/' + account;
+  static std::mutex mutex;
+  std::lock_guard<std::mutex> lock(mutex);
```
consider adding this mutex lock to delete as well.

chandrachivukula replied:
since this could be perf intensive, please also add a comment.
consider making this mutex lock a switch (additional argument, or new method overload) that consumers can use to bypass if they dont use the API in multi threaded environments.

chandrachivukula commented on src/keytar_win.cc:
```diff
@@ -19,6 +20,8 @@ KEYTAR_OP_RESULT SetPassword(const std::string& service,
                  const std::string& password,
                  std::string* errStr) {
   std::string target_name = service + '/' + account;
+  static std::mutex mutex;
```
please add comments why this is necessary for reliable writes.

chandrachivukula reviewed (approved):
approved with comments.

Excellent work digging into the OS layers to uncover the issue and find a workaround, @juturu

husayn committed:
Windows fixes

husayn committed:
update ifdef

husayn committed:
Clean up

huversee committed:
uping version of package json

juturu committed:
Merge pull request #2 from juturu/husayn-keytar4-credoptions

keytar4 add credential options - to support different credential persistance and type

wayvad committed:
Keytar4 (#5)

* Add emebbed typings

* Ignore .tags files and CI configuration in npm

* 4.0.3

* Support unicode account names on windows

* :memo:

[ci-skip]

* Fix typos

* added test to validate the unicode support

* Only build pushes on Travis for master branch

[ci skip]

* force the JS string to use unicode characters

* 4.0.4

* feat: prebuild binaries for electron and node

* add electron 1.8.0

* allow cross-compile for ARM, which requires static linking

* :shirt:

* 4.0.5

* use npm again

* support node 9

* Adds a findCredentials method which yields an array of account/password objects of all matching saved credentials #56

* Fix package-lock.json

* 4.1.0

* update deps

* commit ugly package lock

* Restore .npmignore

* Only upload prebuilds for release tag builds

* Rename auth token variable

* Add Electron 2.0 prebuild

* Update TypeScript typings to latest DefinitelyTyped version

* Add findCredentials signature to keytar.d.ts

* Fix FindCredentials size on windows

`FindCredentials` was missing passing the `CredentialBlobSize` which resulted in an extra character at the end of the password for some values. Fixes: https://github.com/atom/node-keytar/issues/96

* Break long line in src/keytar_win.cc

* Test Unicode support in findCredentials

* FindCredentials should return UTF-8 usernames on Windows

* 4.2.0

* Enable annotated tag builds in appveyor.yml

* Fix missing TypeScript typings in published package

* 4.2.1

* fix typings

* bracket type

* Add N-API support.

* Match the node-addon-api version.

* Update package.json

* Update package.json

* Update README.md

Co-authored-by: Milan Burda <miburda@microsoft.com>
Co-authored-by: Ash Wilson <smashwilson@github.com>
Co-authored-by: Julien Roncaglia <julien@roncaglia.fr>
Co-authored-by: Nguyen Long Nhat <torn4dom4n@gmail.com>
Co-authored-by: Wliu <50Wliu@users.noreply.github.com>
Co-authored-by: Brendan Forster <brendan@github.com>
Co-authored-by: Brendan Forster <github@brendanforster.com>
Co-authored-by: Michelle Tilley <binarymuse@github.com>
Co-authored-by: Vladimir Krivosheev <vladimir.krivosheev@jetbrains.com>
Co-authored-by: develar <develar@gmail.com>
Co-authored-by: Steve Desmond <steve@vtsv.ca>
Co-authored-by: Damien Guard <damien@envytech.co.uk>
Co-authored-by: Damien Guard <damieng@gmail.com>
Co-authored-by: Michael Siebert <michael.siebert@samedi.de>
Co-authored-by: David Wilson <daviwil@github.com>
Co-authored-by: Dotta <34892728+cryppadotta@users.noreply.github.com>
Co-authored-by: Michael Wei <mwei@cs.ucsd.edu>
Co-authored-by: Lee Dohm <1038121+lee-dohm@users.noreply.github.com>
